h2/src/test/org/h2/test/jdbc/TestCustomDataTypesHandler.java

```diff
@@ -0,0 +1,511 @@
+                }
+            }
+            
+            throw DbException.get(
+                    ErrorCode.DATA_CONVERSION_ERROR_1, getString());
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public Value add(Value value) {
+            ValueComplex v = (ValueComplex)value;
+            return ValueComplex.get(val.add(v.val));
+        }
+    }
+
+    /**
+     * Complex number
+     */
+    public static class ComplexNumber implements Serializable {
+        /** */
+        private static final long serialVersionUID = 1L;
+
+        /** */
+        public final static DecimalFormat REAL_FMT = new DecimalFormat("###.###");
+
+        /** */
+        public final static DecimalFormat IMG_FMT = new DecimalFormat("+###.###i;-###.###i");
+
+        /**
+         * Real part
+         */
+        private double re;
+
+        /**
+         * Imaginary part
+         */
+        private double im;
+
+        /**
+         * @param re real part
+         * @param im imaginary part
+         */
+        public ComplexNumber(double re, double im) {
+            this.re = re;
+            this.im = im;
+        }
+
+        /**
+         * Addition
+         * @param other value to add
+         * @return result
+         */
+        public ComplexNumber add(ComplexNumber other) {
+            return new ComplexNumber(re + other.re, im + other.im);
+        }
+
+        /**
+         * Compares two complex numbers
+         *
+         * True ordering of complex number has no sense,
+         * so we apply lexicographical order.
+         *
+         * @param v number to compare this with
+         * @return result of comparison
+         */
+        public int compare(ComplexNumber v) {
+            if (re == v.re && im == v.im) {
+                return 0;
+            }
+            if (re == v.re) {
+                return im > v.im ? 1 : -1;
+            } else if (re > v.re) {
+                return 1;
+            } else {
+                return -1;
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int hashCode() {
+            return (int)re | (int)im;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) {
+                return false;
+            }
+            if (!(other instanceof ComplexNumber)) {
+                return false;
+            }
+            ComplexNumber complex = (ComplexNumber)other;
+            return (re==complex.re) && (im == complex.im);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public String toString() {
+           if (im == 0.0) {
+               return REAL_FMT.format(re);
+           }
+           if (re == 0.0) {
+               return IMG_FMT.format(im);
+           }
+           return REAL_FMT.format(re) + "" + IMG_FMT.format(im);
+        }
```

Comment:
In fact, it will return '1+1i'. If you take a closer look at IMG_FMT declaration, there are two distinct formats for positive and negative number with explicit sign and an 'i' at the very end. I have just specifically test that. Thank you